backtracking.go

```diff
@@ -57,17 +59,27 @@ func (b *Backtracking) Init(f, g float64, step float64) Operation {
 	b.stepSize = step
 	b.initF = f
 	b.initG = g
-	return FuncEvaluation
-}
 
-func (b *Backtracking) Finished(f, _ float64) bool {
-	return ArmijoConditionMet(f, b.initF, b.initG, b.stepSize, b.FuncConst)
+	b.lastOp = FuncEvaluation
+	return b.lastOp
 }
 
-func (b *Backtracking) Iterate(_, _ float64) (float64, Operation, error) {
+func (b *Backtracking) Iterate(f, _ float64) (Operation, float64, error) {
+	if b.lastOp != FuncEvaluation {
+		panic("backtracking: Init has not been called")
+	}
+
+	if ArmijoConditionMet(f, b.initF, b.initG, b.stepSize, b.FuncConst) {
+		return b.next(MajorIteration, nil)
+	}
 	b.stepSize *= b.Decrease
 	if b.stepSize < minimumBacktrackingStepSize {
-		return 0, NoOperation, ErrLinesearchFailure
+		return b.next(NoOperation, ErrLinesearchFailure)
 	}
-	return b.stepSize, FuncEvaluation, nil
+	return b.next(FuncEvaluation, nil)
+}
+
+func (b *Backtracking) next(op Operation, err error) (Operation, float64, error) {
```

Comment:
This function is simple and only happens twice. I think it's better to just put this in line.